nlp_architect/models/tagging.py

```diff
@@ -269,145 +356,28 @@ def train(
-            if epoch_ul > epochs:
-                logger.info("Done")
-                return
-            self.model.train()
-            batch_l, t_batch_l = batch_l[:2]
-            batch_ul, t_batch_ul = batch_ul[:2]
-            t_batch_l = tuple(t.to(self.device) for t in t_batch_l)
-            t_batch_ul = tuple(t.to(self.device) for t in t_batch_ul)
-            t_logits = distiller.get_teacher_logits(t_batch_l)
-            t_logits_ul = distiller.get_teacher_logits(t_batch_ul)
-            batch_l = tuple(t.to(self.device) for t in batch_l)
-            batch_ul = tuple(t.to(self.device) for t in batch_ul)
-            inputs = self.batch_mapper(batch_l)
-            inputs_ul = self.batch_mapper(batch_ul)
-            logits = self.model(**inputs)
-            logits_ul = self.model(**inputs_ul)
-            t_labels = torch.argmax(F.log_softmax(t_logits_ul, dim=2), dim=2)
-            if self.use_crf:
-                loss_labeled = -1.0 * self.crf(logits, inputs["labels"], mask=inputs["mask"] != 0.0)
-                loss_unlabeled = -1.0 * self.crf(logits_ul, t_labels, mask=inputs_ul["mask"] != 0.0)
-            else:
-                loss_fn = CrossEntropyLoss(ignore_index=0)
-                loss_labeled = loss_fn(logits.view(-1, self.num_labels), inputs["labels"].view(-1))
-                loss_unlabeled = loss_fn(logits_ul.view(-1, self.num_labels), t_labels.view(-1))
-
-            if self.n_gpus > 1:
-                loss_labeled = loss_labeled.mean()
-                loss_unlabeled = loss_unlabeled.mean()
-
-            # add distillation loss
-            loss_labeled = distiller.distill_loss(loss_labeled, logits, t_logits)
-            loss_unlabeled = distiller.distill_loss(loss_unlabeled, logits_ul, t_logits_ul)
-
-            # sum labeled and unlabeled losses
-            loss = loss_labeled + loss_unlabeled
-            loss.backward()
-            torch.nn.utils.clip_grad_norm_(self.model.parameters(), max_grad_norm)
-            optimizer.step()
-            # self.model.zero_grad()
-            optimizer.zero_grad()
-            global_step += 1
-            avg_loss += loss.item()
-            if global_step % logging_steps == 0:
-                if s_idx != 0:
-                    logger.info(
-                        " global_step = %s, average loss = %s", global_step, avg_loss / s_idx
-                    )
-                dev = self._get_eval(dev_data_set, "dev")
-                test = self._get_eval(test_data_set, "test")
-                if dev > best_dev:
-                    best_dev = dev
-                    best_test = test
-                    if save_path is not None and save_best:
-                        self.save_model(save_path)
-                logger.info("Best result: dev= %s, test= %s", str(best_dev), str(best_test))
-            if save_path is not None and global_step % save_steps == 0:
-                self.save_model(save_path)
+        self.update_best_model(
```

Comment:
again, `self.update_best_model` shouldn't be a part of the model but as part of the training